File: backend/src/services/tokenBlacklistService.js
import pool from '../config/database.js';
import { AUTH } from '../config/constants.js';

export const cleanupExpiredTokens = async () => {
  try {
    const result = await pool.query(
      `DELETE FROM token_blacklist 
       WHERE created_at < NOW() - INTERVAL '${AUTH.TOKEN_BLACKLIST_CLEANUP_HOURS} hours'`
    );
    console.log(`Cleaned up ${result.rowCount} expired blacklist entries`);
    return result.rowCount;
  } catch (error) {
    console.error('Token cleanup error:', error);
    throw error;
  }
};

export const isTokenBlacklisted = async (userId, tokenIat) => {
  const result = await pool.query(
    `SELECT 1 FROM token_blacklist 
     WHERE user_id = $1 AND created_at > $2 
     LIMIT 1`,
    [userId, new Date(tokenIat * 1000)]
  );
  return result.rows.length > 0;
};

export const addToBlacklist = async (userId) => {
  await pool.query(
    'INSERT INTO token_blacklist (user_id) VALUES ($1)',
    [userId]
  );
};

export const startCleanupSchedule = () => {
  setInterval(cleanupExpiredTokens, 6 * 60 * 60 * 1000);
  cleanupExpiredTokens();
  console.log('Token blacklist cleanup scheduled (every 6 hours)');
};

File: backend/app/services/token_blacklist_service.py
import asyncio
from datetime import datetime, timedelta
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, delete
from app.models.audit import TokenBlacklist
from app.core.logging import logger

TOKEN_BLACKLIST_CLEANUP_HOURS = 168  # 7 days

async def cleanup_expired_tokens(db: AsyncSession) -> int:
    """Delete expired token blacklist entries"""
    try:
        cutoff_time = datetime.utcnow() - timedelta(hours=TOKEN_BLACKLIST_CLEANUP_HOURS)
        result = await db.execute(
            delete(TokenBlacklist).where(TokenBlacklist.created_at < cutoff_time)
        )
        await db.commit()
        count = result.rowcount
        logger.info(f"Cleaned up {count} expired blacklist entries")
        return count
    except Exception as error:
        logger.error(f"Token cleanup error: {error}")
        raise error

async def is_token_blacklisted(db: AsyncSession, user_id: str, token_iat: int) -> bool:
    """Check if token is blacklisted"""
    token_time = datetime.fromtimestamp(token_iat)
    result = await db.execute(
        select(TokenBlacklist)
        .where(TokenBlacklist.user_id == user_id)
        .where(TokenBlacklist.created_at > token_time)
        .limit(1)
    )
    return result.scalar_one_or_none() is not None

async def add_to_blacklist(db: AsyncSession, user_id: str):
    """Add user to token blacklist"""
    blacklist = TokenBlacklist(user_id=user_id)
    db.add(blacklist)
    await db.commit()

async def start_cleanup_schedule(db: AsyncSession):
    """Start periodic cleanup of expired tokens"""
    while True:
        await cleanup_expired_tokens(db)
        await asyncio.sleep(6 * 60 * 60)  # Every 6 hours
